flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java

```diff
@@ -579,6 +580,19 @@ protected static Configuration loadConfiguration(
 
     public static void runClusterEntrypoint(ClusterEntrypoint clusterEntrypoint) {
 
+        // Register the signal handler so that we could have the chance to execute the shutdown
+        // supplier before handling the signal.
+        SignalHandler.register(
+                LOG,
+                () -> {
+                    if (clusterEntrypoint.isShutDown.get()) {
+                        LOG.info("Waiting for concurrent shutDownAsync finished.");
+                        return clusterEntrypoint.terminationFuture.thenAccept(ignored -> {});
+                    } else {
+                        return FutureUtils.completedVoidFuture();
+                    }
+                });
```

Comment:
I am not 100% convinced that we should use the `SignalHandler`. I still like the idea of registering a shutdown hook which simply calls `clusterEntrypoint.closeAsync().join()`. That way we don't have to distinguish between the different signals (SIGTERM and SIGKILL for example).

In order to solve the problem that we don't want to deregister the application if `closeAsync` is called, we could introduce something like a shut down reason. If Flink wants to shut down, then we deregister the application. If somebody else kills the process, then we don't deregister the application. Similar to how we handle the ha cleanup. What do you think?

Reply:
It would also be great if we could guard this behaviour with a test.

Reply:
I think simply calling `clusterEntrypoint.closeAsync().join()` in the shutdown hook could not solve the problem.

When there is a running `shutDownAsync(... cleanupHaData=true)` and we received a concurrent `SIGTERM` signal, the JVM will try to exit and start to execute the shutdown hook. If we only do `clusterEntrypoint.closeAsync()`, we may still have the residual ConfigMaps/ZNodes. Because the `haServices.closeAndCleanupAllData` may not be fully executed.

Inspired by your suggestion, I think we could also wait the possible concurrent `shutDownAsync` finished in `closeAsync()`. After then simply registering the `closeAsync()` as a shutdown hook could work well. It just has same effect with current implementation with `SignalHanlder`.

Reply:
Yeah. What I means is `closeAsync()` needs to return the `terminationFuture` if there was a concurrent `shutDownAsync` already in progress.

I am just saying current `clusterEntrypoint.closeAsync()` not work.

Reply:
I think this is what `closeAsync()` does. It checks whether a shut down is in progress. If not, then it triggers it. If yes, then it simply returns the `terminationFuture`.

Reply:
You are right. I overlooked the `clusterEntrypoint.closeAsync()`.

I am trying to add some tests for this approach and will update this PR soon.